src/sinks/gcp/cloud_storage.rs

```diff
@@ -0,0 +1,550 @@
+    storage_class: Option<GcsStorageClass>,
+    metadata: Option<HashMap<String, String>>,
+    key_prefix: Option<String>,
+    filename_time_format: Option<String>,
+    filename_append_uuid: Option<bool>,
+    filename_extension: Option<String>,
+    encoding: Encoding,
+    compression: Compression,
+    #[serde(default)]
+    batch: BatchBytesConfig,
+    #[serde(default)]
+    request: TowerRequestConfig,
+    #[serde(flatten)]
+    auth: GcpAuthConfig,
+    tls: Option<TlsOptions>,
+}
+
+#[derive(Clone, Copy, Debug, Derivative, Deserialize, Serialize)]
+#[derivative(Default)]
+#[serde(rename_all = "camelCase")]
+enum GcsPredefinedAcl {
+    AuthenticatedRead,
+    BucketOwnerFullControl,
+    BucketOwnerRead,
+    Private,
+    #[derivative(Default)]
+    ProjectPrivate,
+    PublicRead,
+}
+
+#[derive(Clone, Copy, Debug, Derivative, Deserialize, Serialize)]
+#[derivative(Default)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+enum GcsStorageClass {
+    #[derivative(Default)]
+    Standard,
+    Nearline,
+    Coldline,
+    Archive,
+}
+
+lazy_static! {
+    static ref REQUEST_DEFAULTS: TowerRequestConfig = TowerRequestConfig {
+        in_flight_limit: Some(25),
+        rate_limit_num: Some(25),
+        ..Default::default()
+    };
+}
+
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Copy, Derivative)]
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+enum Encoding {
+    #[derivative(Default)]
+    Text,
+    Ndjson,
+}
+
+impl Encoding {
+    fn content_type(&self) -> &'static str {
+        match self {
+            Self::Text => "text/plain",
+            Self::Ndjson => "application/x-ndjson",
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, Copy, Derivative)]
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+enum Compression {
+    #[derivative(Default)]
+    Gzip,
+    None,
+}
+
+impl Compression {
+    fn content_encoding(&self) -> Option<&'static str> {
+        match self {
+            Self::Gzip => Some("gzip"),
+            Self::None => None,
+        }
+    }
+    fn extension(&self) -> &'static str {
+        match self {
+            Self::Gzip => "log.gz",
+            Self::None => "log",
+        }
+    }
+}
+
+inventory::submit! {
+    SinkDescription::new::<GcsSinkConfig>(NAME)
+}
+
+#[typetag::serde(name = "gcp_cloud_storage")]
+impl SinkConfig for GcsSinkConfig {
+    fn build(&self, cx: SinkContext) -> crate::Result<(RouterSink, Healthcheck)> {
+        let sink = GcsSink::new(self, &cx)?;
+        let healthcheck = sink.healthcheck()?;
+        let service = sink.service(self, &cx)?;
+
+        Ok((service, healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        NAME
+    }
+}
+
+#[derive(Debug, Snafu)]
+enum HealthcheckError {
+    #[snafu(display("Invalid credentials"))]
+    InvalidCredentials,
+    #[snafu(display("Unknown bucket: {:?}", bucket))]
```

Comment:
Seems like this is ok to be a normal display `{}` instead of a debug `{:?}`.

Reply:
I normally do things like this as a debug, as it puts quotes around it and escapes characters, to make obvious if there was something like an extraneous space in the config, or unusual characters, etc.